File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {

private:
    vector<string> res;
public:
    vector<string> generateParenthesis(int n) {

        backtrack(0, 0, "", n);
        return res;
        
    }

    void backtrack(int open, int close, string str, int n) {
        if (open == n && close == n) {
            res.push_back(str);
            return;
        }

        if (open < n) {
            str += "(";
            backtrack(open + 1, close, str, n);
            str.pop_back();
        }

        if (close < open) {
            str += ")";
            backtrack(open, close + 1, str, n);
            str.pop_back();
        }
    }


};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        self.res = []
        self.n = n

        openBrackets = 0
        closeBrackets = 0

        self.DFS("", openBrackets, closeBrackets)
        return self.res

    def DFS(self, substr, openBrackets, closeBrackets):
        if openBrackets == self.n and closeBrackets == self.n:
            self.res.append(substr)
            return
            
        if openBrackets < self.n:
            openBrackets += 1
            self.DFS(substr + '(', openBrackets, closeBrackets)
            openBrackets -= 1
            
        if closeBrackets < openBrackets:
            closeBrackets += 1
            self.DFS(substr + ')', openBrackets, closeBrackets)
            closeBrackets -= 1

            




        